File: scripts/import_to_supabase.py
"""
import_to_supabase.py
=====================
Importa os CSVs da migração Feegow diretamente no Supabase via API REST.

Pré-requisito:
    pip install requests

Uso:
    python import_to_supabase.py

Os CSVs devem estar na pasta migration_output/ (mesma pasta deste script).
"""

import csv
import json
import os
import time
from pathlib import Path

try:
    import requests
except ImportError:
    print("❌ Instale o requests primeiro:  pip install requests")
    raise

# ---------------------------------------------------------------------------
# CONFIGURAÇÃO — não precisa alterar nada
# ---------------------------------------------------------------------------
SUPABASE_URL = "https://iudghjdmfcxjcpspeify.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Iml1ZGdoamRtZmN4amNwc3BlaWZ5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjgzMDI5NTQsImV4cCI6MjA4Mzg3ODk1NH0.f2rcdjPuxyx2REEviDRObE-Sn1D4ytb4_HwTDjW8Xao"

HEADERS = {
    "apikey":        SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type":  "application/json",
    "Prefer":        "return=minimal",  # mais rápido: não retorna os dados inseridos
}

SCRIPT_DIR = Path(__file__).parent
CSV_DIR    = SCRIPT_DIR / "migration_output"

# Ordem importa: professionals e services antes de patients, patients antes de appointments
IMPORT_ORDER = [
    ("professionals", "professionals.csv"),
    ("services",      "services.csv"),
    ("patients",      "patients.csv"),
    ("appointments",  "appointments.csv"),
]

BATCH_SIZE = 500   # número de linhas por request (Supabase suporta até ~1000)

# ---------------------------------------------------------------------------

def load_csv(path: Path) -> list[dict]:
    with open(path, encoding="utf-8") as f:
        return list(csv.DictReader(f))


def clean_row(row: dict) -> dict:
    """Converte strings vazias para None e garante tipos corretos."""
    cleaned = {}
    for k, v in row.items():
        if v == "" or v is None:
            cleaned[k] = None
        else:
            cleaned[k] = v
    return cleaned


def insert_batch(table: str, rows: list[dict]) -> tuple[bool, str]:
    """Insere um lote de linhas. Retorna (sucesso, mensagem_erro)."""
    url = f"{SUPABASE_URL}/rest/v1/{table}"
    payload = [clean_row(r) for r in rows]
    try:
        r = requests.post(url, headers=HEADERS, data=json.dumps(payload), timeout=30)
        if r.status_code in (200, 201):
            return True, ""
        else:
            return False, f"HTTP {r.status_code}: {r.text[:200]}"
    except requests.exceptions.RequestException as e:
        return False, str(e)


def import_table(table: str, csv_file: Path):
    print(f"\n  Importando {table}...")
    if not csv_file.exists():
        print(f"    ⚠️  Arquivo não encontrado: {csv_file}")
        return

    rows = load_csv(csv_file)
    total   = len(rows)
    success = 0
    errors  = 0

    for i in range(0, total, BATCH_SIZE):
        batch = rows[i : i + BATCH_SIZE]
        ok, err = insert_batch(table, batch)
        if ok:
            success += len(batch)
            pct = int(success / total * 100)
            print(f"    {pct:3d}% — {success}/{total}", end="\r")
        else:
            errors += len(batch)
            print(f"\n    ❌ Erro no lote {i}-{i+len(batch)}: {err}")

        time.sleep(0.1)  # evita rate limiting

    print(f"    ✅ {table}: {success} inseridos, {errors} erros          ")


def test_connection():
    print("🔌 Testando conexão com Supabase...")
    try:
        r = requests.get(
            f"{SUPABASE_URL}/rest/v1/professionals?limit=1",
            headers=HEADERS,
            timeout=10
        )
        if r.status_code == 200:
            print("  ✅ Conexão OK!")
            return True
        else:
            print(f"  ❌ Erro de conexão: HTTP {r.status_code}")
            print(f"     {r.text[:300]}")
            return False
    except Exception as e:
        print(f"  ❌ Não foi possível conectar: {e}")
        return False


def main():
    print(f"\n{'='*56}")
    print(f"  Importação Feegow → Supabase (SintesIA)")
    print(f"{'='*56}")

    if not test_connection():
        print("\n  Verifique a chave do Supabase e tente novamente.")
        return

    print(f"\n📁 Pasta dos CSVs: {CSV_DIR}")

    for table, csv_filename in IMPORT_ORDER:
        import_table(table, CSV_DIR / csv_filename)

    print(f"\n{'='*56}")
    print(f"  ✅ Importação concluída!")
    print(f"     Verifique no Supabase Dashboard → Table Editor")
    print(f"{'='*56}\n")


if __name__ == "__main__":
    main()

File: scripts/import_to_supabase.mjs
/**
 * import_to_supabase.mjs — Importa dados Feegow → Supabase
 * Uso: node import_to_supabase.mjs
 */

import { readFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const SUPABASE_URL = "https://iudghjdmfcxjcpspeify.supabase.co";
const SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Iml1ZGdoamRtZmN4amNwc3BlaWZ5Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc2ODMwMjk1NCwiZXhwIjoyMDgzODc4OTU0fQ.cnD-3_9kfasLNi8W9WMlGmRIJj9JOH8H8wDqP-HFYCc";

const JSON_DIR   = join(__dirname, "migration_output");
const BATCH_SIZE = 300;

const HEADERS = {
  "apikey":        SUPABASE_KEY,
  "Authorization": `Bearer ${SUPABASE_KEY}`,
  "Content-Type":  "application/json",
  "Prefer":        "resolution=merge-duplicates,return=minimal", // upsert seguro
};

const IMPORT_ORDER = [
  { table: "professionals", file: "professionals.json" },
  { table: "services",      file: "services.json"      },
  { table: "patients",      file: "patients.json"      },
  { table: "appointments",  file: "appointments.json"  },
];

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

function chunks(arr, size) {
  const out = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

async function upsertBatch(table, rows) {
  const url = `${SUPABASE_URL}/rest/v1/${table}`;
  const res = await fetch(url, {
    method:  "POST",
    headers: HEADERS,
    body:    JSON.stringify(rows),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`HTTP ${res.status}: ${body.slice(0, 250)}`);
  }
}

async function testConnection() {
  process.stdout.write("🔌 Testando conexão... ");
  const res = await fetch(`${SUPABASE_URL}/rest/v1/professionals?limit=1`, {
    headers: { "apikey": SUPABASE_KEY, "Authorization": `Bearer ${SUPABASE_KEY}` }
  });
  if (res.ok) { console.log("✅ OK!"); return true; }
  console.log(`❌ Falhou (HTTP ${res.status})`);
  return false;
}

async function importTable(table, jsonFile) {
  const path = join(JSON_DIR, jsonFile);
  console.log(`\n  📥 ${table}...`);

  let rows;
  try { rows = JSON.parse(readFileSync(path, "utf-8")); }
  catch (e) { console.log(`    ⚠️  Arquivo não encontrado: ${path}`); return; }

  const batches = chunks(rows, BATCH_SIZE);
  let done = 0, errors = 0;

  for (const batch of batches) {
    try {
      await upsertBatch(table, batch);
      done += batch.length;
    } catch (e) {
      errors += batch.length;
      console.log(`\n    ❌ Erro no lote: ${e.message}`);
    }
    process.stdout.write(`    ${Math.round(done / rows.length * 100)}% — ${done}/${rows.length}\r`);
    await sleep(100);
  }

  const status = errors === 0 ? "✅" : "⚠️ ";
  console.log(`    ${status} ${done} inseridos${errors ? `, ${errors} erros` : ""}            `);
}

async function main() {
  console.log("\n" + "=".repeat(52));
  console.log("  Importação Feegow → Supabase (SintesIA)");
  console.log("=".repeat(52));

  if (!await testConnection()) { process.exit(1); }

  for (const { table, file } of IMPORT_ORDER) {
    await importTable(table, file);
  }

  console.log("\n" + "=".repeat(52));
  console.log("  ✅ Pronto! Confira em supabase.com → Table Editor");
  console.log("=".repeat(52) + "\n");
}

main().catch(e => { console.error("Erro:", e); process.exit(1); });
